### ISSUE TITLE
Persona login window shows "emails" and "success" rather than the emails addresses to select

### ISSUE THREAD

**Issue by Standard8** [mozilla/persona#3010]
STR:

1) I've signed in with persona on one site (e.g. etherpad).
2) I go to log into a different site

Expected Results:

Show me the dialog with the two email addresses that I've got listed.

Actual Results:
 I get the normal dialog, however, the radio buttons to select my two email addresses are "emails" and "success" rather than the actual email address.

You can see an example here:

![Screen Shot 2013-02-14 at 13 49 28](https://f.cloud.github.com/assets/520844/156939/25eb5fee-76af-11e2-8446-e0d9480e59f6.png)

The only possible relevant item on the error console was:

Timestamp: 14/02/2013 14:00:26
Warning: WARN addons.manager: Exception calling callback: [Exception... "Component returned failure code: 0x804b000a (NS_ERROR_MALFORMED_URI) [nsIIOService.newURI]"  nsresult: "0x804b000a (NS_ERROR_MALFORMED_URI)"  location: "JS frame :: resource://gre/modules/NetUtil.jsm :: NetUtil_newURI :: line 198"  data: no]
Source File: resource://gre/modules/NetUtil.jsm
Line: (null)

On the site I was logged into last night, I had to log out fully, then mess around a lot, and eventually was able to log back in again.

However, it appears to have yet again reverted to this state.

If I visit https://login.persona.org/ then I also see "emails" and "success".

**Comment by ozten** [comment]
@Standard8 These are secondary email addresses, right? (You have to enter a password into Persona... it isn't backed by a different identity provider).

**Comment by ozten** [comment]
I'm unable to reproduce this issue. I did see Issue #3011, which seems to not reproduce in a clean browser profile.

Can you
1) Email me the two email addresses you are using.
2) Try to reproduce in a new browser profile
3) Using Firebug or another tool, [save out a HAR](http://www.softwareishard.com/blog/har-adopters/) file of signing in

2 and 3 are bonus points :)

**Comment by Standard8** [comment]
No, one should be my primary, the other my secondary. I see this on the main persona site (that has remembered my password).

**Comment by lloyd** [comment]
cache inconsistency?  strange that it resolves and then resurfaces.

Can you reproduce this on any other browser or device?

**Comment by callahad** [comment]
I would also like to see the contents of localStorage for the login.persona.org domain...

**Comment by callahad** [comment]
(Which you can get via `JSON.stringify(window.localStorage)` in the web console when you're on login.persona.org)

Please make sure to remove / mangle any "`priv`" keys if they show up in the output.

**Comment by ozten** [comment]
@Standard8 What browser version and OS version are you on?

Lowering from 5 stars, as I cannot reproduce and we need more information to debug.

**Comment by wfwalker** [comment]
I experience this problem now with FF 21.0a1 (2013-02-06) on Mac OS X 10.8.2 trying to log into apps.etherpad.mozilla.org
But on the very same machine with FF 18.0.2 (using a different profile), all is well.

**Comment by ozten** [comment]
These two reports also mention etherpad which uses the old getVerifiedEmail API.

**Comment by jrgm** [comment]
So, you can artificially repro this, like so:
1. Spin up an awsbox for train-2012.11.23
2. Signup as a user with the awsbox (get the email, etc. ...)
3. ssh to the box and save ~app/code/resources/static/production/en/dialog.js to ~app
4. Now force push train-2013.02.01 to the awsbox
5. On the box cp ~app/dialog.js ~app/code/resources/static/production/en/dialog.js
6. Bring up the dialog
You now see 'success' and 'emails' as choices of identity.

**Comment by ozten** [comment]
@jrgm Awesome work!

Prod fix in PR #3013

We have discussed three ways to fix the root issue

1) Add a version [mismatch detection](https://github.com/mozilla/browserid/pull/2934)
2) Change deployment techniques
3) Making forwards and backwards compatible components
4) Making connect-cachify respond with a 404 when the request has a wrong SHA
4a) Make this a configuration that is changeable at runtime

**Comment by lloyd** [comment]
How did we get in this state?  Code was pushed to a web head and software was not restarted?

**Comment by ozten** [comment]
@lloyd the hypothesis is a browser with stale DNS information talking to the "drained" data center. A server running old code responded to the new url with old code.

**Comment by lloyd** [comment]
Given the widespread impact, are we satisfied with this hypothesis?  

Don't we wait DNs TTL before updating?

Given frequency I'd think its more likely that we have a static process that hasn't been restarted.

Recall, one reporter flushed cache, got
Fixed, the got back into this state.

Can we confirm all static processes have been restarted?

-- lloyd (thumb-typing)

**Comment by lloyd** [comment]
strikes me that we could request the resource in question 20x per dc and md5 response to verify we're serving consistent resources.

**Comment by fmarier** [comment]
@lloyd In practice, DNS TTLs are just a suggestion. I know some ISPs that configure their resolvers to always cache for at least 1 day.

For example, I moved my blog over to a new server about a month ago and I still see people hitting the old IP. For some reason, they're almost all running Opera...

**Comment by ozten** [comment]
> strikes me that we could request the resource in question 20x per dc and md5 response to verify we're serving consistent resources.

I didn't catalog every detail of troubleshooting.

jrgm and I independently both had checked that... used curl and md5 and found no variance. jrgm being more thorough did it 100 times... I had only done it about 30 ...

**Comment by jrgm** [comment]
> Given the widespread impact, are we satisfied with this hypothesis?

Yeah, it's my hypothesis and makes sense for a particular case, but doesn't fit all the data perfectly (as you note below).

> Don't we wait DNs TTL before updating?

We do. TTL is 30 seconds and it's several minutes before updates happen.

**Comment by lloyd** [comment]
@ozten / @jrgm both DCs?

@francois I know, but if it were DNS related that we would see it every deployment.  we don't.  So this explanation is unsatisfying.  Especially because the initial reporter had it occur, had it fixed, then _had it occur again_

Sorry to hammer on this, but I really think a deep understanding of root cause would be useful.  Thanks everyone for digging into this on valentines day.

**Comment by jrgm** [comment]
@lloyd both DCs, by changing /etc/hosts (and checking with netstat that I am actually using the IP I expect).

**Comment by lloyd** [comment]
I guess I've just rehashed the work @jrgm and @ozten did yesterday.  I apologize, as I'm not really convinced I added any value.

FWIW, The suggestion of @ozten above of "Making connect-cachify respond with a 404 when the request has a wrong SHA" sounds right on - as the root cause of cache poisioning here was static serving resources that did not match sha inside the url.  If we do this we turn cache poisioning (blocks user for 2ish weeks) intto a broken dialog (blocks user for a couple minutes).

The final bit I cannot explain is why the impact hit 2 users who actually spoke up (implying greater impacT).  The best theory here is that switchover from PHX1 to SJC at 2013-02-13 15:36:36 PST was not atomic, and users got `/sign_in` from SCL2 and then dialog.js from PHX1.  That vulnerability window is very small if everyone is respecting TTL.  @jrgm suggests not everyone does.
